### DISCUSSION TITLE
Support previous session in callbacks

### DISCUSSION THREAD

**Discussion by lavcraft** [nextauthjs/next-auth#3946] [Ideas]
### Description 📓

Provide current session variable in callbacks for access information stored in session and merge into new.

It's useful in cases, when you need to get some tokens from multiple service for provide integration. For example — you need to create documents in sharepoint and attach it to google calendars. In this case, i need in one time and one place two tokens — from google and from azure.

This ability in callbacks is very useful for cover case above, and accumulate all user accounts (with different emails in place, when i confident about this merge (but if it main flow for authenticated user, like "attach google account" - why not?)

According to [Discussion about handling current session](https://github.com/nextauthjs/next-auth/discussions/2976#discussioncomment-1488293) — it's not so hard to implement feature :)

### How to reproduce ☕️

For examples usage (see current* args):

```javascript
        jwt: async (
            newToken: JWT,
            user?: User,
            account?: Account,
            profile?: Profile,
            isNewUser?: boolean,
            currentToken?: JWT
        )  = { ... }
```

Or

```javascript
session: async (
            newSession: Session,
            userOrToken: JWT | User,
            oldSession: Session,
        ) => {..}
```

### Contributing 🙌🏽

Yes, I am willing to help implement this feature in a PR

**Comment by dbeaudway** [comment]
@lavcraft, I'm very interested in this feature and have seen other discussion posts asking for the same. Any progress here -- willing to pitch in if you have a branch.

**Comment by balazsorban44** [comment]
Related: https://github.com/nextauthjs/next-auth/discussions/3116#discussioncomment-1622935

**Comment by lavcraft** [comment]
@balazsorban44 thanks. It's related, but not similar one. In this case, we don't need multiple sessions info in stateless token. Only need current session during authenticating process for merging.

For example:
1. Login via google (email1)
2. Login via azure (email2) while authorized from step 1. In signIn or session callback i can resolve current session and construct new session with derived roles.  Also - can write new email (email2 from azure) to user info in database (may spawn additional dialog for user, like do you want attach this account to your user?)

**Comment by rnnyrk** [comment]
Would be really helpful to have more control over the session with the option for a previous session. Currently working on an Auth0 implementation where MFA requires a different token. Doing any kind of new login via next auth replaces the old session, where it would be ideal to save the MFA session along with the current session.

**Comment by raphaelpc** [comment]
Hey! Any update on this? :)

**Reply by lavcraft** [reply]
Sorry, but no. Can't wait for start implement this feature, but not enough time for this :( Not so important for my business, currently procrastinating on this pr

**Comment by balazsorban44** [comment]
Progress on this https://github.com/nextauthjs/next-auth/pull/6337#issuecomment-1376490423
